### ISSUE TITLE
Visual Studio Exception and hundreds of requests when adding react-bootstrap

### ISSUE THREAD

**Issue by mayerwin** [ServiceStack/ServiceStackVS#38]
I am trying to use `react-bootstrap` with the TypeScript+React template. I installed it with:

```
jspm install react-bootstrap
typings install dt~react-bootstrap --global --save
```

However running `00-update-deps-js` will not help hundreds of requests to be made (due to the many components in react-bootstrap), and Visual Studio does report the following exception in the immediate window (not reported as a Javascript error though):
```
Exception was thrown at line 6155, column 3 in eval code
0x800a01b6 - JavaScript runtime error: Object doesn't support property or method 'Symbol(Symbol.iterator)_3.zc0te5lu9g8'
```

You  can reproduce by simply replacing the app.tsx code with:

```
/// <reference path='../typings/index.d.ts'/>

import * as React from 'react';
import * as ReactDOM from 'react-dom';
import { Jumbotron } from 'react-bootstrap';

export default class Home extends React.Component<any, any> {
    constructor(props, context) {
        super(props, context);
    }

    render() {
        return (
            <Jumbotron>
                <h1>Welcome</h1>
                <p>Let''s do some interesting stuff today!</p>
                {/*<p><Button bsStyle="primary">Learn more</Button></p>*/}
            </Jumbotron>);
    }
}

ReactDOM.render(<Home />, document.getElementById('content'));
```

This is the minimum reproducible error I am able to find. My project was not working at all in IE11 (which provides useful Javascript debugging from the IDE itself), so I tried to understand what was going on. Solving this error may solve all the other cryptic errors I was encountering, or at least give a starting point.

One thing I was wondering: what is the purpose of `deps.tsx`? Should all dependencies be listed there as imports (note: adding react-bootstrap there did not help)? I thought jspm was able to automatically and smartly do the equivalent of webpack if we provide him with the entry point (`app.tsx`), but it seems the gulp task is looking at `deps.js` instead.

Another side question: is it possible to enable hot module reloading? Normally JSPM supports that (in beta).

**Comment by mythz** [comment]
The hundreds of requests is exactly the problem with npm's preference for micro modules and why we've added `deps.tsx` to try help minimize the number of requests. `deps.tsx` is something we've added, it's not apart of JSPM, but we use jspm's bundle feature in the `00-update-deps-js` to build a new cache. 

You need to manually update `deps.tsx` to include references to your new components so they're included as part of the cache. When JSPM loads it will make requests for any dependencies that's not already loaded (i.e. that's not in deps.lib.js). 

> I thought jspm was able to automatically and smartly do the equivalent of webpack if we provide him with the entry point (app.tsx), but it seems the gulp task is looking at deps.js instead.

JSPM is doing it's job, but if we didn't have `deps.tsx` to prime it's initial module cache, every module your app depends on would need to be loaded on every request, which is very slow for the hundreds of requests an empty React App needs. So to minimize the requests, after adding any major components you should add a reference to something from that module in `deps.tsx` and then re-run update-deps to re-generate the `deps.lib.js` cache.

**Comment by mayerwin** [comment]
I was able to do away with the `deps.tsx` file with the following modified gulp task inspired from [systemjs doc](https://github.com/systemjs/builder#bundle-arithmetic):

    gulp.task('www-jspm-deps', function () {
        return gulp.src('./src/app.js')
            .pipe(jspmBuild({ arithmetic: '- [./src/**/*] - [src/app/**/*!text]' }))
            .pipe(rename('deps.lib.js'))
            .pipe(gulp.dest('./'));
    });
	
According to [this post](http://code.rawlinson.us/2015/11/jspm-bundling-options.html), `- [src/app/**/*!text]` is needed as well to exclude any html file (I wonder why as * is supposed to be a wildcard, isn't it?).
If you confirm this is a good idea, it would be worth updating your templates to make use of this.

> Note: we only bundle the App at deployment, at runtime JSPM is just loading the generated .js files that VS/TypeScript generates on Save as it provides the fastest iterative development cycle.

Does this mean we can refresh the browser after saving a modified `*.tsx` file and it should just use the latest version? It'd be great if the templates could be updated, once JSPM 0.17 is out, to take advantage of [hot reloading](http://jspm.io/0.17-beta-guide/hot-reloading.html).

With regard to 0.17, it also adds a built-in watcher and development bundler to solve the [issue](https://github.com/jspm/jspm-cli/issues/872): 

> jspm 0.17-0-beta.12 now supports jspm bundle x --watch and jspm build x --watch commands.

It would make a real difference in terms of iteration speed to see [these capabilities](http://jspm.io/0.17-beta-guide/development-bundling.html
) implemented in the gulp file:

> we will use the jspm bundle command with the watch flag:
> 
> jspm bundle test.js -wid
> where the flags here are:
> 
> --watch: Watch the module tree for file changes to then do a quick cached rebuild.
> --inject: Ensure that the bundle is automatically loaded without any further changes necessary.
> --development: Create a development bundle, without any production flags or optimizations, which are otherwise the default when creating a bundle.
> 
> Now any changes to any of the modules will automatically rebuild the bundle, refreshing the browser will then load the changes from that single bundle file for a quick page refresh.

**Comment by mythz** [comment]
ok it looks like this is useful:

```js
    gulp.task('www-jspm-deps', function () {
        return gulp.src('./src/app.js')
            .pipe(jspmBuild({ arithmetic: '- [./src/**/*]' }))
            .pipe(rename('deps.lib.js'))
            .pipe(gulp.dest('./'));
    });
```

I had an error trying to run the original:

```ts
.pipe(jspmBuild({ arithmetic: '- [./src/**/*] - [src/app/**/*!text]' }))
```

Which failed with:

```
gulp-jspm: Error: (SystemJS) ENOENT: no such file or directory, open 'C:\src\wip\ReactTSApp1\ReactTSApp1\text.js'
	Error: ENOENT: no such file or directory, open 'C:\src\wip\ReactTSApp1\ReactTSApp1\text.js'
	    at Error (native)
	Error loading C:/src/wip/ReactTSApp1/ReactTSApp1/text.js
	    at Error (native)
```

So I'll look at updating the template with this change to replace `deps.tsx`.

> Does this mean we can refresh the browser after saving a modified *.tsx file and it should just use the latest version? 

Yes that's a feature built into TypeScript's support for VS.NET which will autogenerate `.js` files after saving each `.ts` or `.tsx` file. It also means Users don't need to run any Gulp Tasks during their normal edit/save/reload dev cycle.

Which also means there's no dependency on a background watcher processes to monitor file changes and regenerate TypeScript .js files. We want to make the templates as simple to understand with the least moving parts as possible which IMO the current template maintains a good balance of and why we don't want to enable any watchers or hot reloading by default. If preferred these could be local customizations added after for users who prefer this functionality.

But we don't think the current template is the ultimate solution and are keen to follow solutions that work best with TypeScript which unfortunately hasn't always worked well with React were its TypeScript declarations often get out of sync with React's libraries. Hopefully future releases of TypeScript includes a bundling and dependency management solution which we can use instead of JSPM as we'd expect an official solution with Q/A would cause the least issues.

**Comment by mythz** [comment]
I've made a few tweaks to the SPA Templates to replace deps.tsx with your www-jspm-deps update solution, simplified the gulpfile tasks, added `import "es6-shim"` to app.tsx and also copying over `system-polyfills.js` so the default template should now work by default in IE11. 

Please try it out and let me know if you run into any issues?

[ServiceStackVS.vsix](http://ci.servicestack.net:2080/repository/download/ServiceStackVS_Build/7320:id/ServiceStackVS.vsix)

You'll need to uninstall **ServiceStackVS** VS.NET Extension from `Tools > Extensions and Updates` before installing this one.

**Comment by mayerwin** [comment]
Great. Happy to test, however your link seems to be password-protected.

**Comment by mythz** [comment]
Apologies that was our internal CI url, we also publish to a public S3 at:

https://s3.amazonaws.com/ci-dist/ssvs/latest/ServiceStackVS.vsix

**Comment by mayerwin** [comment]
Looks good to me (I just saw your reply), except in `gulp.js` you seem to have forgotten to replace `www-clean-client-assets` with `www-clean-client` and `www-clean-dlls `with `www-clean-server` at a few places.
I wish Visual Studio would not report the exceptions caught (and properly handled) from IE, but there is most likely nothing we can do about it.

**Comment by mythz** [comment]
> except in `gulp.js` you seem to have forgotten to replace `www-clean-client-assets` with `www-clean-client` and `www-clean-dlls `with `www-clean-server` at a few places.

Which places do you mean where they were not replaced?

**Comment by mayerwin** [comment]
Sorry, it seems I checked in an outdated file, your latest version seems fine.

**Comment by mythz** [comment]
FYI I've made a [number of improvements to the TypeScript React App](https://github.com/ServiceStack/ServiceStackVS/commit/7271e10ca9e981036e24fbe1c36eb248dc81064e) Template if you want to test it out:

- Removed typings, using `@types` or devDependencies for Type Defs
- Upgraded to jspm@beta
- Optimized bundling to stand-alone app.js with jspm build + rollup
- Upgraded to bootstrap@v4
- Use camelCase by default
- Removed deps.tsx + automatically generating deps.js on npm install
- Upgraded tsconfig.json to use built-in es2015 type defs + removed
custom fileGlob

Basically I've simplified it as much as possible where I've switched to using built-in type defs or newer `@types` when not included in npm package. Using the latest version of jspm now allows us to build an optimized stand-alone app.js which doesn't require the jspm runtime and merges all dep registrations into a single module which you can checkout by viewing the source of a deployed default template at: http://vstest.servicestack.net/

Latest version of ServiceStackVS is available at: https://s3.amazonaws.com/ci-dist/ssvs/latest/ServiceStackVS.vsix

**Comment by mayerwin** [comment]
Just curious, did you take a look at the latest official templates for [.NET React SPA](https://blogs.msdn.microsoft.com/webdev/2017/02/14/building-single-page-applications-on-asp-net-core-with-javascriptservices/)? 

They look rather interesting have support the following:
- Server-side prerendering
- Webpack dev middleware
- Hot Module Replacement
- Efficient production builds

Obviously they are not as comprehensive as ServiceStackVS (there is no packaging into cross platform apps), but it may be beneficial to draw inspiration from it for the ultimate developer experience.

**Comment by mythz** [comment]
We've already ditched JSPM and moved to Webpack for our React App VS.NET Templates in our dev build of SSVS. Unfortunately NodeServices and Webpack Middlware are only available in ASP.NET Core Apps so we wont be able to offer server-side pre-rendering but we do have gulp tasks and npm scripts for running `webpack watch` which auto recompiles TypeScript behind the scenes as well as a live test runner for TypeScript tests using Facebook's jest. The [WebpackDevMiddleware](https://github.com/aspnet/JavaScriptServices/blob/dev/src/Microsoft.AspNetCore.SpaServices/Webpack/WebpackDevMiddleware.cs#L86) impl looks a bit ugly, it spawns a node.js embedded `webpack-dev-middleware` server that it [proxies all requests to](https://github.com/aspnet/JavaScriptServices/blob/dev/src/Microsoft.AspNetCore.SpaServices/Webpack/ConditionalProxyMiddleware.cs#L51), I can look into taking a similar approach to get it working in ASP.NET 4.5.

For .NET Core templates we're going to have to create new SPA templates from scratch, we'll look at seeing if we can use Microsoft's SPA templates as a base add any additional value that's missing, but CefSharp isn't available for .NET Core so we wouldn't to offer a React Desktop Apps template.

**Comment by mayerwin** [comment]
Thanks for the news, great to know, I'll check out the latest version when it is released and let you know if I have any comment. Exciting times for .NET for sure.
